jfix-stdlib-ratelimiter/src/main/kotlin/ru/fix/stdlib/ratelimiter/SuspendableRateLimitedDispatcher.kt

```diff
@@ -0,0 +1,282 @@
+     * and should not block processing thread.
+     *
+     * @param supplier task to execute and retrieve result
+     * @return feature which represent result of task execution
+     */
+    private suspend fun <T> submit(supplier: () -> T): Deferred<T> {
+        val state = state.get()
+        return if (state != State.RUNNING) {
+            throw RejectedExecutionException(
+                    "RateLimiterDispatcher [$name] is in '$state' state"
+            )
+        } else {
+            val queueWaitTime = profiler.start("queue_wait")
+            queueSize.incrementAndGet()
+
+            val deferred = coroutineScope.async {
+                processTask(Task2(supplier, queueWaitTime))
+            }
+            deferreds.add(deferred)
+            deferred.invokeOnCompletion { deferreds.remove(deferred) }
+            deferred
+        }
+    }
+
+    fun updateRate(rate: Int) {
+        rateLimiter.updateRate(rate)
+    }
+
+    override fun close() {
+        var stateUpdated = state.compareAndSet(State.RUNNING, State.SHUTTING_DOWN)
+        if (!stateUpdated) {
+            logger.info("Close called on RateLimitedDispatcher [{}] with state [{}]", name, state.get())
+            return
+        }
+        windowSizeSubscription.close()
+
+        deferreds.forEach {
+            it.cancel("Cancelling deferred on dispatcher close")
+        }
+
+        if (closingTimeout.get() < 0) {
+            logger.warn(
+                "Rate limiter timeout must be greater than or equals 0. Current value is {}, rate limiter name: {}",
+                closingTimeout.get(), name
+            )
+        }
+        stateUpdated = state.compareAndSet(State.SHUTTING_DOWN, State.TERMINATE)
+        if (!stateUpdated) {
+            logger.error(
+                "Can't set [TERMINATE] state to RateLimitedDispatcher [{}] in [{}] state",
+                name, state.get()
+            )
+            return
+        }
+        rateLimiter.close()
+        profiler.detachIndicator(QUEUE_SIZE_INDICATOR)
+        profiler.detachIndicator(ACTIVE_ASYNC_OPERATIONS)
+    }
+
+
+    private suspend fun <T> processTask(task: Task2<T>): T {
+        task.queueWaitTimeCall.stop()
+        try {
+            if (windowSize > 0) {
+                profiler.start("acquire_window").use { acquireWindowTime ->
+                    var windowAcquired = false
+                    while (!windowAcquired) {
+                        if (state.get() == State.TERMINATE) {
+                            val message = "RateLimitedDispatcher [$name] received TERMINATE state while window acquiring"
+                            logger.error(message)
+                            throw RejectedExecutionException(message)
+                        }
+                        windowAcquired = semaphore.get().tryAcquire()
+                        if (!windowAcquired) {
+                            delay(WINDOW_ACQUIRE_DELAY_MS)
+                        }
```

Comment:
In such way we can spent extra time on window wait. The proper way of acquiring it in coroutines without cpu resource consumption on delays is:
```suggestion
                        windowAcquired = withTimeoutOrNull(3_000) {
                            semaphore.get().acquire()
                            true
                        } ?: false
```

Reply:
Ok